objects/src/accounts/delta/vault.rs

```diff
@@ -1,236 +1,488 @@
+            .into_iter()
+            .map(|(account_id, delta_as_u64)| (account_id, delta_as_u64 as i64))
+            .collect();
+
+        Self::new(map).map_err(|err| DeserializationError::InvalidValue(err.to_string()))
+    }
+}
+
+/// A binary tree map of non-fungible asset changes (addition and removal) in the account vault.
+#[derive(Clone, Debug, Default, PartialEq, Eq)]
+pub struct NonFungibleAssetDelta(BTreeMap<Digest, NonFungibleDeltaAction>);
+
+impl NonFungibleAssetDelta {
+    /// Validates and creates a new non-fungible asset delta.
+    ///
+    /// # Errors
+    /// Returns an error if the delta does not pass the validation.
+    pub fn new(map: BTreeMap<Digest, NonFungibleDeltaAction>) -> Result<Self, AccountDeltaError> {
+        let delta = Self(map);
         delta.validate()?;
 
         Ok(delta)
     }
 
-    /// Checks whether this vault delta is valid.
+    /// Adds a new non-fungible asset to the delta.
     ///
     /// # Errors
-    /// Returns an error if:
-    /// - The number added assets is greater than [u16::MAX].
-    /// - The number of removed assets is greater than [u16::MAX].
-    /// - The same asset was added more than once, removed more than once, or both added and
-    ///   removed.
-    pub fn validate(&self) -> Result<(), AccountDeltaError> {
-        if self.added_assets.len() > u16::MAX as usize {
-            return Err(AccountDeltaError::TooManyAddedAsset {
-                actual: self.added_assets.len(),
-                max: u16::MAX as usize,
-            });
-        } else if self.removed_assets.len() > u16::MAX as usize {
-            return Err(AccountDeltaError::TooManyRemovedAssets {
-                actual: self.removed_assets.len(),
-                max: u16::MAX as usize,
-            });
-        }
+    /// Returns an error if the delta already contains the asset addition.
+    pub fn add(&mut self, asset: NonFungibleAsset) -> Result<(), AccountDeltaError> {
+        self.apply_action(Word::from(asset).into(), NonFungibleDeltaAction::Add)
+    }
 
-        // make sure all added assets are unique
-        for (pos, asset) in self.added_assets.iter().enumerate() {
-            if self.added_assets[..pos].iter().any(|a| a.is_same(asset)) {
-                return Err(AccountDeltaError::DuplicateVaultUpdate(*asset));
-            }
+    /// Removes a non-fungible asset from the delta.
+    ///
+    /// # Errors
+    /// Returns an error if the delta already contains the asset removal.
+    pub fn remove(&mut self, asset: NonFungibleAsset) -> Result<(), AccountDeltaError> {
+        self.apply_action(Word::from(asset).into(), NonFungibleDeltaAction::Remove)
+    }
+
+    /// Returns true if this vault delta contains no updates.
+    pub fn is_empty(&self) -> bool {
+        self.0.is_empty()
+    }
+
+    /// Returns an iterator over the (key, value) pairs of the map.
+    pub fn iter(&self) -> impl Iterator<Item = (&Digest, &NonFungibleDeltaAction)> {
```

Comment:
If the previous comment works, then the return type here can be `impl Iterator<Item = (&NonFungibleAsset, &NonFungibleDeltaAction)>`